File: python/snake.py
# coding: CP852

import os, sys, random, pygame
from pygame.locals import *

White  = (0xFF, 0xFF, 0xFF)
Red    = (0xFF, 0x00, 0x00)
Yellow = (0xFF, 0xFF, 0x00)
Green  = (0x00, 0x80, 0x00)
Azure  = (0x40, 0x80, 0xFF)
Blue   = (0x00, 0x00, 0x80)
Black  = (0x00, 0x00, 0x00)
BKG    = Green

class Surface(pygame.Surface):

	def drawCell(self, margin, color, fill = True):

		pygame.draw.rect(self, color, (margin, margin, Cell - 2 * margin, Cell - 2 * margin), not fill)

class Snake:

	def __init__(self):

		self.head = [(Width // 2) * Cell, (Height - 1) * Cell]
		self.body = []

		self.himg = Surface(CellSize)
		self.bimg = Surface(CellSize)

		self.himg.fill(Blue)
		self.bimg.fill(BKG)
		self.bimg.drawCell(1, Azure)
		self.bimg.drawCell(4, Blue)

		self.length = 3
		self.angle  = 0
		self.alive  = True

	def turn_left(self):

		self.angle = (self.angle + 3) % 4

	def turn_right(self):

		self.angle = (self.angle + 1) % 4

	def move(self):

		self.body.insert(0, self.head)
		self.body = self.body[:self.length]
		self.head = self.head[:]
		if   self.angle == 0: self.head[1] -= Cell
		elif self.angle == 1: self.head[0] += Cell
		elif self.angle == 2: self.head[1] += Cell
		elif self.angle == 3: self.head[0] -= Cell

		if self.head in self.body or self.head[0] < 0 or self.head[0] >= ScreenWidth or self.head[1] < 0 or self.head[1] >= ScreenHeight:
			self.alive = False

	def update(self):

		screen = pygame.display.get_surface()
		screen.blit(self.himg, self.head)
		for b in self.body: screen.blit(self.bimg, b)

class Fruit:

	def __init__(self):

		self.bkg = Surface(CellSize); self.bkg.fill(BKG)
		self.bkg.drawCell(1, Yellow, False)
		self.img = Surface(CellSize); self.img.fill(BKG)
		self.img.drawCell(1, Yellow, False)
		self.img.drawCell(4, Red)
		self.move()

	def move(self):

		self.score = 9
		self.ticks = pygame.time.get_ticks() + 1000
		self.pos = [random.randint(0, Width - 1) * Cell, random.randint(0, Height - 1) * Cell]

	def update(self):

		if self.score > 0 and pygame.time.get_ticks() >= self.ticks: self.score -= 1; self.ticks += 1000
		screen = pygame.display.get_surface()
		screen.blit(self.img if pygame.time.get_ticks() % 1024 < 512 else self.bkg, self.pos)

def showInfo(s, l):

	screen = pygame.display.get_surface()
	font = pygame.font.SysFont("Consolas", 24)
#	font = pygame.font.Font(None, 32)

	text = font.render(str(s), 1, White)
	rect = text.get_rect()
	rect = rect.move(4, 4)
	screen.blit(text, rect)

	text = font.render(str(l), 1, White)
	rect = text.get_rect()
	rect = rect.move(ScreenWidth - rect.right - 4, 4)
	screen.blit(text, rect)

def showEnd():

	screen = pygame.display.get_surface()
	font = pygame.font.SysFont("Consolas", 48)
#	font = pygame.font.Font(None, 64)

	text = font.render("GAME OVER", 1, White)
	rect = text.get_rect(); rect.center = ScreenWidth // 2, ScreenHeight // 2
	screen.blit(text, rect)

	pygame.display.update()
	pygame.time.wait(2000)

def main():
    
	screen = pygame.display.set_mode(ScreenSize)
	snake  = Snake()
	fruit  = Fruit()

	while snake.alive:

		for e in pygame.event.get():
			if e.type == QUIT or e.type == KEYDOWN and e.key == K_ESCAPE: return
			if e.type == KEYDOWN:
				if e.key == K_LEFT:  snake.turn_left()
				if e.key == K_RIGHT: snake.turn_right()
		snake.move()
		if fruit.pos in snake.body: snake.length += fruit.score; fruit.move()
		screen.fill(BKG)
		snake.update()
		fruit.update()
		showInfo(fruit.score, len(snake.body))
		pygame.display.update()
		pygame.time.wait(100)

	showEnd()

CellSize = Cell, Cell = 16, 16
Size = Width, Height = 40, 30
ScreenSize = ScreenWidth, ScreenHeight = Width * Cell, Height * Cell

if __name__ == "__main__":

	os.environ["SDL_VIDEO_CENTERED"] = "1"
#	os.environ["SDL_VIDEODRIVER"]    = "windib"
	pygame.init()
	pygame.display.set_caption("SNAKE!")
	main()
	pygame.quit()

File: ruby/snake.rb
require 'gosu'

White  = Gosu::Color.argb(0xFF_FFFFFF)
Red    = Gosu::Color.argb(0xFF_FF0000)
Yellow = Gosu::Color.argb(0xFF_FFFF00)
Green  = Gosu::Color.argb(0xFF_008000)
Azure  = Gosu::Color.argb(0xFF_4080FF)
Blue   = Gosu::Color.argb(0xFF_000080)
Black  = Gosu::Color.argb(0xFF_000000)
BKG    = Green

module ZOrder
	BACKGROUND, SNAKE, FRUIT, TEXT = *0..3
end

def draw_cell(x, y, margin, color, zorder)
	Gosu::draw_rect(x + margin, y + margin, Cell - 2 * margin, Cell - 2 * margin, color, zorder)
end

class Array
	def out_of_size?(size)
		each_with_index { |x, i| return true if x < 0 or x >= size[i] }
		false
	end
end

class Snake

	def initialize
		@head = [(Width / 2) * Cell, (Height - 1) * Cell]
		@body = []

		@length = 3
		@angle  = 0
		@alive  = true
	end

	def turn_left
		@angle = (@angle + 3) % 4
	end

	def turn_right
		@angle = (@angle + 1) % 4
	end

	def move
		@body.unshift(@head.dup)
		@body.pop if @body.length > @length
#		@body = @body[0, @length]

		case @angle
		when 0 then @head[1] -= Cell
		when 1 then @head[0] += Cell
		when 2 then @head[1] += Cell
		when 3 then @head[0] -= Cell
		end

		@alive = false if @body.include?(@head) or @head.out_of_size?(ScreenSize)
	end

	def update
		move
		if gulp?($fruit)
			rise
			$fruit.create
		end
	end
	
	def draw
		draw_cell(*@head, 0, Blue,  ZOrder::SNAKE)
		@body.each do |b|
			draw_cell(*b, 0, BKG,   ZOrder::SNAKE)
			draw_cell(*b, 1, Azure, ZOrder::SNAKE)
			draw_cell(*b, 4, Blue,  ZOrder::SNAKE)
		end
	end

	def gulp?(obj)
		@head == obj.pos
	end

	def size
		@body.length
	end

	def rise
		@length += $score
	end

	def alive?
		@alive
	end

end

class Fruit

	attr_reader :pos

	def initialize
		create
	end

	def create
		@pos   = [rand(Width) * Cell, rand(Height) * Cell]
		$score = 9
		@timer = Gosu::milliseconds + 1000
	end

	def update
		if Gosu::milliseconds >= @timer
			$score -= 1 if not $score.zero?
			@timer += 1000
		end
	end
	
	def draw
		draw_cell(*@pos, 0, BKG,    ZOrder::FRUIT)
		draw_cell(*@pos, 1, Yellow, ZOrder::FRUIT)
		c = Red
		c.alpha = (Gosu::milliseconds / 4) % 256
		draw_cell(*@pos, 4, c, ZOrder::FRUIT)
	end

end

class Game < Gosu::Window
	
	def initialize
		super(*ScreenSize, false, 100)
		self.caption = 'SNAKE!'
		@font = Gosu::Font.new(24)
		@large = Gosu::Font.new(48)
	end

	def button_down(id)
		close if id == Gosu::KbEscape
		$snake.turn_left  if id == Gosu::KbLeft
		$snake.turn_right if id == Gosu::KbRight
	end

	def update
		if $snake.alive?
			$fruit.update
			$snake.update
			@stop = Gosu::milliseconds + 2000
		end
		close if Gosu::milliseconds >= @stop
	end

	def draw
		Gosu::draw_rect(*ScreenRect, BKG, ZOrder::BACKGROUND)
		$snake.draw
		$fruit.draw
		showInfo
		showEnd if $snake.alive? == false
	end

	private

	def showInfo
		@font.draw_text_rel("#{$score}", 4, 4, ZOrder::TEXT, 0.0, 0.0)
		@font.draw_text_rel("#{$snake.size}", ScreenWidth - 4, 4, ZOrder::TEXT, 1.0, 0.0)
	end

	def showEnd
		@large.draw_text_rel("GAME OVER", ScreenWidth / 2, ScreenHeight / 2, ZOrder::TEXT, 0.5, 0.5)
	end

end

Cell = 16
Size = (Width, Height = 40, 30)
ScreenSize = (ScreenWidth, ScreenHeight = Width * Cell, Height * Cell)
ScreenRect = 0, 0, *ScreenSize

if __FILE__ == $0

	$score = 0
	$snake = Snake.new
	$fruit = Fruit.new
	$game  = Game.new
	$game.show

end
